presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java

```diff
@@ -74,18 +93,32 @@
             .with(asyncTimeout)
             .with(Executors.newCachedThreadPool(daemonThreadsNamed("docker-container-%d")));
 
+    private final FixedPortsAdapter portsAdapter = new FixedPortsAdapter();
     private final String logicalName;
-    private final Stopwatch startupTime = Stopwatch.createUnstarted();
+    private final EnvironmentListenerAdapter listenerAdapter;
+    private final GenericContainer<? extends GenericContainer> delegate;
+
     private List<String> logPaths = new ArrayList<>();
-    private Optional<EnvironmentListener> listener = Optional.empty();
 
     public DockerContainer(String dockerImageName, String logicalName)
     {
-        super(dockerImageName);
+        this(logicalName, fixedHostPort(dockerImageName), container -> {});
+    }
+
+    public <T extends GenericContainer<T>> DockerContainer(String logicalName, DelegateContainerFactory<T> containerFactory, Consumer<T> containerClosure)
+    {
         this.logicalName = requireNonNull(logicalName, "logicalName is null");
+        requireNonNull(containerFactory, "containerFactory is null");
+        requireNonNull(containerClosure, "containerFactory is null");
```

Comment:
should be `containerClosure` in error message

Reply:
https://gist.github.com/findepi/b597d2bc91e5a30c2a44fd65fe252f0b